CRM/Core/BAO/CustomGroup.php

```diff
@@ -28,6 +28,83 @@ public static function self_hook_civicrm_post(\Civi\Core\Event\PostEvent $e): vo
     Civi::cache('metadata')->flush();
   }
 
+  /**
+   * Master function to retrieve custom groups with their fields
+   *
+   * This provides a cached list of all custom groups, with a nested array of fields.
+   *
+   * @param bool $includeDisabled
+   *   If false (default) this will filter out disabled groups and fields.
+   * @return array[]
+   */
+  public static function getAll(bool $includeDisabled = FALSE): array {
+    $cacheString = __CLASS__ . __FUNCTION__ . '_' . CRM_Core_I18n::getLocale();
+    $allGroups = Civi::cache('metadata')->get($cacheString);
+    if (!is_array($allGroups)) {
+      $allGroups = self::loadAll();
+      Civi::cache('metadata')->set($cacheString, $allGroups);
+    }
+    if (!$includeDisabled) {
+      $allGroups = array_values(array_filter($allGroups, fn($group) => $group['is_active']));
+      foreach ($allGroups as $groupIndex => $group) {
+        $allGroups[$groupIndex]['fields'] = array_values(array_filter($group['fields'], fn($field) => $field['is_active']));
+      }
+    }
+    return $allGroups;
+  }
+
+  /**
+   * Same as self::getAll() but returns only active groups and
+   * fields which the given user is allowed to see.
+   *
+   * @param int $permissionType (CRM_Core_Permission::VIEW or CRM_Core_Permission::EDIT)
+   * @param int|null $userId
+   * @return array[]
+   */
+  public static function getPermitted(int $permissionType, ?int $userId = NULL) {
+    if (!in_array($permissionType, [CRM_Core_Permission::EDIT, CRM_Core_Permission::VIEW], TRUE)) {
+      throw new CRM_Core_Exception('permissionType must be CRM_Core_Permission::VIEW or CRM_Core_Permission::EDIT');
+    }
+    $allGroups = self::getAll();
+    $allowedGroupIds = CRM_Core_Permission::customGroup($permissionType, FALSE, $userId);
+    $allGroups = array_filter($allGroups, function($group) use ($allowedGroupIds) {
+      return in_array($group['id'], $allowedGroupIds);
+    });
+    return array_values($allGroups);
+  }
+
+  /**
+   * Loads all custom groups and fields.
+   *
+   * Avoids calling the API to prevent recursion or early-bootstrap issues.
+   *
+   * @return array
+   */
+  private static function loadAll(): array {
+    $custom = [];
```

Comment:
doesn't the caching belong to this function, not `getAll()`

Reply:
Yea can do. Then it can be a public function.

Reply:
yeah - I'm on the fence about making it public - just wondering a use case for that